sklearn/cluster/_k_means_lloyd.pyx

```diff
@@ -0,0 +1,398 @@
+        Centers before previous iteration, placeholder for the centers after
+        previous iteration.
+
+    centers_new : {float32, float64} array-like, shape (n_clusters, n_features)
+        Centers after previous iteration, placeholder for the new centers
+        computed during this iteration.
+
+    centers_squared_norms : {float32, float64} array-like, shape (n_clusters,)
+        Squared L2 norm of the centers.
+
+    weight_in_clusters : {float32, float64} array-like, shape (n_clusters,)
+        Placeholder for the sums of the weights of every observation assigned
+        to each center.
+
+    labels : int array-like, shape (n_samples,)
+        labels assignment.
+
+    center_shift : {float32, float64} array-like, shape (n_clusters,)
+        Distance between old and new centers.
+
+    n_threads : int
+        The number of threads to be used by openmp.
+
+    update_centers : bool
+        - If True, the labels and the new centers will be computed, i.e. runs
+          the E-step and the M-step of the algorithm.
+        - If False, only the labels will be computed, i.e runs the E-step of
+          the algorithm. This is useful especially when calling predict on a
+          fitted model.
+    """
+    cdef:
+        int n_samples = X.shape[0]
+        int n_features = X.shape[1]
+        int n_clusters = centers_new.shape[0]
+
+        # hard-coded number of samples per chunk. Appeared to be close to
+        # optimal in all situations.
+        int n_samples_chunk = 256 if n_samples > 256 else n_samples
+        int n_chunks = n_samples // n_samples_chunk
+        int n_samples_rem = n_samples % n_samples_chunk
+        int chunk_idx, n_samples_chunk_eff
+        int start, end
+
+        int j, k
+
+        floating[::1] centers_squared_norms = row_norms(centers_old, squared=True)
+
+        floating *centers_new_chunk
+        floating *weight_in_clusters_chunk
+        floating *pairwise_distances_chunk
+
+    # count remainder chunk in total number of chunks
+    n_chunks += n_samples != n_chunks * n_samples_chunk
+
+    if update_centers:
+        memset(&centers_new[0, 0], 0, n_clusters * n_features * sizeof(floating))
+        memset(&weight_in_clusters[0], 0, n_clusters * sizeof(floating))
+
+    with nogil, parallel(num_threads=n_threads):
+        # thread local buffers
+        centers_new_chunk = <floating*> calloc(n_clusters * n_features, sizeof(floating))
+        weight_in_clusters_chunk = <floating*> calloc(n_clusters, sizeof(floating))
+        pairwise_distances_chunk = <floating*> malloc(n_samples_chunk * n_clusters * sizeof(floating))
+
+        for chunk_idx in prange(n_chunks):
+            start = chunk_idx * n_samples_chunk
+            if chunk_idx == n_chunks - 1 and n_samples_rem > 0:
+                end = start + n_samples_rem
+            else:
+                end = start + n_samples_chunk
+
+            _update_chunk_dense(
```

Comment:
Hum I don't if I can do something about that. It seems related to openmp and error handling but I don't really understand the generated code. However I don't think it's really an issue because each step of the loop does a lot of computations.

The interaction at the end of the parallel section is definitely not an issue since it's outside of the loop

Reply:
> Hum I don't if I can do something about that. It seems related to openmp and error handling but I don't really understand the generated code. 

Yeah... I had a similar issue in the HGBT code, and for reference the reason I found was that I was passing down slices of 2d arrays to a function, i.e. `X[:, feature_idx]`. So I changed everything to accept 2d views instead of 1d views. 

> However I don't think it's really an issue because each step of the loop does a lot of computations

If there are lots of samples (and thus chunks) that might add up. But agreed, probably not worth the pain...

Reply:
Here, in terms of slices, I'm only passing 1D slices of 1D arrays, but it might be the same issue.
I guess passing pointers would solve the issue as well but it would make the code harder to follow so I'd rather not do that.